src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceUpdateService.java

```diff
@@ -0,0 +1,181 @@
+import org.opensearch.geospatial.ip2geo.common.Ip2GeoSettings;
+
+@Log4j2
+public class DatasourceUpdateService {
+    private final ClusterService clusterService;
+    private final Client client;
+    private final DatasourceHelper datasourceHelper;
+    private final GeoIpDataHelper geoIpDataHelper;
+    private TimeValue timeout;
+    private Integer indexingBulkSize;
+
+    public DatasourceUpdateService(
+        final ClusterService clusterService,
+        final Client client,
+        final DatasourceHelper datasourceHelper,
+        final GeoIpDataHelper geoIpDataHelper
+    ) {
+        this.clusterService = clusterService;
+        this.client = client;
+        this.datasourceHelper = datasourceHelper;
+        this.geoIpDataHelper = geoIpDataHelper;
+
+        this.timeout = Ip2GeoSettings.TIMEOUT_IN_SECONDS.get(clusterService.getSettings());
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(Ip2GeoSettings.TIMEOUT_IN_SECONDS, newValue -> this.timeout = newValue);
+        this.indexingBulkSize = Ip2GeoSettings.INDEXING_BULK_SIZE.get(clusterService.getSettings());
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, newValue -> this.indexingBulkSize = newValue);
+    }
+
+    /**
+     * Update GeoIp data internal
+     *
+     * @param jobParameter
+     * @throws Exception
+     */
+    public void updateGeoIpData(final Datasource jobParameter) throws Exception {
+        if (!DatasourceState.AVAILABLE.equals(jobParameter.getState())) {
+            log.error("Invalid datasource state. Expecting {} but received {}", DatasourceState.AVAILABLE, jobParameter.getState());
+            jobParameter.disable();
+            jobParameter.getUpdateStats().setLastFailedAt(Instant.now());
+            datasourceHelper.updateDatasource(jobParameter);
+            return;
+        }
+
+        URL url = new URL(jobParameter.getEndpoint());
+        DatasourceManifest manifest = DatasourceManifest.Builder.build(url);
+
+        if (skipUpdate(jobParameter, manifest)) {
+            log.info("Skipping GeoIP database update. Update is not required for {}", jobParameter.getId());
+            jobParameter.getUpdateStats().setLastSkippedAt(Instant.now());
+            datasourceHelper.updateDatasource(jobParameter);
+            return;
+        }
+
+        Instant startTime = Instant.now();
+        String indexName = jobParameter.indexNameFor(manifest);
+        jobParameter.getIndices().add(indexName);
+        datasourceHelper.updateDatasource(jobParameter);
+        geoIpDataHelper.createIndexIfNotExists(indexName);
+        String[] fields;
+        List<String> fieldsToStore;
+        try (CSVParser reader = geoIpDataHelper.getDatabaseReader(manifest)) {
+            Iterator<CSVRecord> iter = reader.iterator();
+            fields = iter.next().values();
+            if (fields.length < 2) {
```

Comment:
you are assuming that header and records length will always be equal. I think you have to validate every record to verify whether they have at least two fields

Reply:
Added validation code